framework/src/utils/DelimitedFileReader.C

```diff
@@ -0,0 +1,184 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+// STL includes
+#include <sstream>
+#include <iomanip>
+
+// MOOSE includes
+#include "DelimitedFileReader.h"
+#include "MooseUtils.h"
+#include "MooseError.h"
+
+namespace MooseUtils
+{
+
+DelimitedFileReader::DelimitedFileReader(const std::string & filename,
+                                         const bool header,
+                                         const std::string delimiter,
+                                         std::shared_ptr<libMesh::Parallel::Communicator> comm)
+  : _filename(filename), _header(header), _delimiter(delimiter), _communicator(comm)
+{
+}
+
+void
+DelimitedFileReader::read()
+{
+  // Number of columns
+  std::size_t n_cols;
+
+  // Storage for the raw data
+  std::vector<double> raw;
+  std::size_t size_raw;
+
+  // Read data
+  if (_communicator->rank() == 0)
+  {
+    // Check the file
+    MooseUtils::checkFileReadable(_filename);
+
+    // Create the file stream
+    std::ifstream stream_data(_filename);
+
+    // Read the header and data
+    initializeColumns(stream_data);
+
+    // Set the number of columns
+    n_cols = _column_names.size();
+
+    // Read the header and data
+    readData(stream_data, raw);
+
+    // Close the stream
+    stream_data.close();
+
+    // Set raw data vector size
+    size_raw = raw.size();
+  }
+
+  // Broadcast column names
+  _communicator->broadcast(n_cols);
```

Comment:
I like the NULL idea, like this...

```
if (_cummunicator == nullptr || _cummincator.rank() == 0)
```

And then the broadcast stuff wrapped in:

```
if (_cummunicator != nullptr)
```